### DISCUSSION TITLE
Is the installation tool required or can Shlink be setup with just env variables?

### DISCUSSION THREAD

**Discussion by msbt** [shlinkio/shlink#1721] [Help wanted] [answered]
#### How Shlink is set up

* Shlink Version: 3.5.2
* PHP Version: 8.1
* How do you serve Shlink: Self-hosting, trying to create a package for Cloudron.io
* Database engine used: MySQL

#### Summary

Hi there! Quick question about the Shlink setup: The [installation tool page in the docs](https://shlink.io/documentation/command-line-interface/installation-tool/) says

> Shlink provides a helper CLI installation tool which comes in handy to configure your instance when you are not using Shlink's docker image, and you prefer not to use [environment variables](https://shlink.io/documentation/environment-variables/).

I would prefer to use environment variables and there are several steps which are missing in the docs. 

With just the env vars, the db never got initialized, I solved that by running `bin/cli db:create` and/or `bin/cli db:migrate`.

When accessing the created URLs I get a 500 server error and the logs said `[2023-03-12T02:12:00.971351+00:00] [37f5e414-c981-4864-89a6-b105a1717213] Shlink.ERROR - ErrorException: require(data/proxies/__CG__ShlinkioShlinkRestEntityApiKey.php): Failed to open stream: No such file or directory in /app/data/vendor/doctrine/common/src/Proxy/AbstractProxyFactory.php:197`

This I could solve by running `php vendor/doctrine/orm/bin/doctrine.php orm:generate-proxies`

Since I never ran `vendor/bin/shlink-installer install` there is also no `config/params/generated_config.php`, will this be a problem at some point?

Am I missing something else which is not created by only using env vars and not running the installation tool?

---

AUTHOR EDIT: See the answer here https://github.com/shlinkio/shlink/discussions/1721#discussioncomment-7025367

**Comment by acelaya** [comment]
The intention is that you can use either the installation tool, or environment variables. The fact that `generated_config.php` does not exist in the second case should not be a problem, as long as the config is provided via env vars.

That said, it is true that there's a gap in the docs on how to properly support this use case.

### Some context

On the very beginning of the project, the installation tool was the only way to set-up Shlink. A couple of years after, the official docker image was introduced, and support for env vars was added. Also, an [entry point script](https://github.com/shlinkio/shlink/blob/658303d3754b172113563148356ff0bff239e1be/docker/docker-entrypoint.sh) was created, which covered some of the extra tasks performed by the installation tool.

At first it was not possible to use env vars for non-docker setups, but I realized it would be useful to support them, allowing for better automation of Shlink deployments on non-container contexts.

However, I made some assumptions here by mistake, and as you figured out, if the installation tool is never invoked, there are a couple missing critical steps that make Shlink not work. You mostly figured out everything (I'm impressed, BTW 😅).

### Next steps

I'm not sure what's the best approach here. I could just document this to begin with, but I feel Shlink should provide some kind of tooling, or be smarter about this. I need to give it a thought.

Also, this is related https://github.com/shlinkio/shlink-installer/issues/176

**Reply by msbt** [reply]
Thanks for the quick reply and clarification @acelaya! 

Quick follow-up on that one: [Cloudron](https://www.cloudron.io/) uses custom Docker images for the available applications where everything is read-only for security reasons, except the files that need to be editable, they'll be part of the automatic backups, the rest is not. If I set Shlink up with environment variables, is the /data folder the only one that needs write permissions?

**Reply by acelaya** [reply]
> If I set Shlink up with environment variables, is the /data folder the only one that needs write permissions?

Yes, that's the only folder where Shlink might need to write something.

**Reply by msbt** [reply]
Ok I've made some progress, but after initialization I can't access the links or get any replies because of 

```
[php:error] PHP Fatal error: Uncaught ArgumentCountError: Too few arguments to function Shlinkio\\Shlink\\Config\\parseEnvVar(), 0 passed in /app/code/vendor/shlinkio/shlink-config/functions/functions.php on line 82 and exactly 1 expected in /app/code/vendor/shlinkio/shlink-config/functions/functions.php:46
Stack trace:
#0 /app/code/vendor/shlinkio/shlink-config/functions/functions.php(82): Shlinkio\\Shlink\\Config\\parseEnvVar()
#1 /app/code/config/autoload/swoole.global.php(26): Shlinkio\\Shlink\\Config\\getOpenswooleConfigFromEnv()
#2 /app/code/config/autoload/swoole.global.php(34): Laminas\\ConfigAggregator\\PhpFileProvider::{closure}()
#3 /app/code/vendor/laminas/laminas-config-aggregator/src/PhpFileProvider.php(29): include('...')
#4 /app/code/vendor/laminas/laminas-config-aggregator/src/ConfigAggregator.php(238): Laminas\\ConfigAggregator\\PhpFileProvider->__invoke()
#5 /app/code/vendor/laminas/laminas-config-aggregator/src/ConfigAggregator.php(82): Laminas\\ConfigAggregator\\ConfigAggregator->loadConfigFromProviders()
#6 /app/code/config/config.php(52): Laminas\\ConfigAggregator\\ConfigAggregator->__construct()
#7 /app/code/config/container.php(26): require('...')
#8 /app/code/config/container.php(31): {closure}()
#9 /app/code/config/run.php(10): include('...')
#10 /app/code/public/index.php(5): {closure}()
#11 {main}
thrown in /app/code/vendor/shlinkio/shlink-config/functions/functions.php on line 46
```

Any idea what that's about? I added most of the environment variables, put them in a .sh file and had it `export envvar=...` which then gets `source env.sh`'ed on every start.

Everything except the data directory resides in `/app/code`, `/app/code/data` is symlinked to `/app/data/data` (yep, that's on purpose). Anything I'm missing?

**Reply by acelaya** [reply]
You are not the first one reporting this, so it might be an actual bug, not necessarily related with your original question.

See https://github.com/shlinkio/shlink/issues/1722

**Reply by acelaya** [reply]
In order to reproduce it, could you run this PHP script in the context you are trying to run Shlink?

```php
<?php
var_dump(getenv());
```

Basically you can replace/comment out the contents of `/app/code/public/index.php` with the above.

**Reply by msbt** [reply]
I left all the other vars in there and sorted them alphabetically, it seems to me that there's an extra space added to every env var set for Shlink at the end, I'll investigate why that is:

```
array(62) { 
  ["ANONYMIZE_REMOTE_ADDR"]=> string(5) "true " 
  ["APACHE_LOCK_DIR"]=> string(17) "/var/lock/apache2" 
  ["APACHE_LOG_DIR"]=> string(16) "/var/log/apache2" 
  ["APACHE_PID_FILE"]=> string(28) "/var/run/apache2/apache2.pid" 
  ["APACHE_RUN_DIR"]=> string(16) "/var/run/apache2" 
  ["APACHE_RUN_GROUP"]=> string(8) "www-data" 
  ["APACHE_RUN_USER"]=> string(8) "www-data" 
  ["AUTO_RESOLVE_TITLES"]=> string(6) "false " 
  ["CLOUDRON"]=> string(1) "1" 
  ["CLOUDRON_API_ORIGIN"]=> string(16) "https://my.example.com" 
  ["CLOUDRON_APP_DOMAIN"]=> string(12) "slinky.example.com" 
  ["CLOUDRON_APP_HOSTNAME"]=> string(36) "redacted" 
  ["CLOUDRON_APP_ORIGIN"]=> string(20) "https://slinky.example.com" 
  ["CLOUDRON_MYSQL_DATABASE"]=> string(16) "redacted" 
  ["CLOUDRON_MYSQL_HOST"]=> string(5) "mysql" 
  ["CLOUDRON_MYSQL_PASSWORD"]=> string(48) "redacted" 
  ["CLOUDRON_MYSQL_PORT"]=> string(4) "3306" 
  ["CLOUDRON_MYSQL_URL"]=> string(96) "mysql://redacted:redacted@mysql/redacted" 
  ["CLOUDRON_MYSQL_USERNAME"]=> string(16) "redacted" 
  ["CLOUDRON_PROXY_IP"]=> string(10) "172.18.0.1" 
  ["CLOUDRON_WEBADMIN_ORIGIN"]=> string(16) "https://my.example.com" 
  ["DB_DRIVER"]=> string(6) "mysql " 
  ["DB_HOST"]=> string(6) "mysql " 
  ["DB_NAME"]=> string(17) "redacted " 
  ["DB_PASSWORD"]=> string(49) "redacted " 
  ["DB_PORT"]=> string(5) "3306 " 
  ["DB_USER"]=> string(17) "redacted " 
  ["DEBIAN_FRONTEND"]=> string(14) "noninteractive" 
  ["DEFAULT_BASE_URL_REDIRECT"]=> string(16) "https://redacted.com " 
  ["DEFAULT_DOMAIN"]=> string(13) "slinky.example.com " 
  ["DEFAULT_INVALID_SHORT_URL_REDIRECT"]=> string(16) "https://redacted.com " 
  ["DEFAULT_QR_CODE_ERROR_CORRECTION"]=> string(2) "l " 
  ["DEFAULT_QR_CODE_FORMAT"]=> string(4) "png " 
  ["DEFAULT_QR_CODE_MARGIN"]=> string(2) "0 " 
  ["DEFAULT_QR_CODE_ROUND_BLOCK_SIZE"]=> string(5) "true " 
  ["DEFAULT_QR_CODE_SIZE"]=> string(4) "300 " 
  ["DEFAULT_REGULAR_404_REDIRECT"]=> string(16) "https://redacted.com " 
  ["DEFAULT_SHORT_CODES_LENGTH"]=> string(2) "5 " 
  ["DISABLE_IP_TRACKING"]=> string(6) "false " 
  ["DISABLE_REFERRER_TRACKING"]=> string(6) "false " 
  ["DISABLE_TRACKING"]=> string(6) "false " 
  ["DISABLE_TRACKING_FROM"]=> string(11) "172.18.*.* " 
  ["DISABLE_UA_TRACKING"]=> string(6) "false " 
  ["GEOLITE_LICENSE_KEY"]=> string(1) " " 
  ["GOROOT"]=> string(20) "/usr/local/go-1.19.3" 
  ["HOSTNAME"]=> string(36) "redacted" 
  ["IS_HTTPS_ENABLED"]=> string(5) "true " 
  ["LANG"]=> string(1) "C" 
  ["MERCURE_INTERNAL_HUB_URL"]=> string(1) " " 
  ["MERCURE_JWT_SECRET"]=> string(1) " " 
  ["MERCURE_PUBLIC_HUB_URL"]=> string(1) " " 
  ["MULTI_SEGMENT_SLUGS_ENABLED"]=> string(6) "false " 
  ["PATH"]=> string(113) "/usr/local/node-18.12.1/bin:/usr/local/go-1.19.3/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" 
  ["PWD"]=> string(9) "/app/code" 
  ["REDIRECT_APPEND_EXTRA_PATH"]=> string(6) "false " 
  ["REDIRECT_CACHE_LIFETIME"]=> string(3) "30 " 
  ["REDIRECT_STATUS_CODE"]=> string(4) "302 " 
  ["SHLVL"]=> string(1) "0" 
  ["SHORT_URL_MODE"]=> string(7) "strict " 
  ["SHORT_URL_TRAILING_SLASH"]=> string(6) "false " 
  ["TERM"]=> string(5) "xterm" 
  ["TRACK_ORPHAN_VISITS"]=> string(5) "true " 
} 
```

The env.sh looks like this:
```
#!/bin/bash
# general
export DEFAULT_DOMAIN="${CLOUDRON_APP_DOMAIN}"
export IS_HTTPS_ENABLED=true
export GEOLITE_LICENSE_KEY=
export AUTO_RESOLVE_TITLES=false
export DEFAULT_SHORT_CODES_LENGTH=5
#export DELETE_SHORT_URL_THRESHOLD=
...
```

**Reply by acelaya** [reply]
I have just tried to run the function throwing that error with that list of env vars, and it does not error out 🤔 

In fact it's very weird. What the logic does is get the list of env vars, filter out those which key does not start with `OPENSWOOLE_` (which in your case it's all of them), and then walk the result and call another function with every value.

On paper, there should be no way for the function to be called without params, because it should effectively be called only if there's a param.

> it seems to me that there's an extra space added to every env var set for Shlink at the end

This should not be the cause to this specific error, but it would be good to make sure those spaces are not there, at least for the boolean values to be properly parsed.

I will put some mitigation in place, but still better to make sure the values Shlink gets are the expected ones.

Perhaps, make sure to quote all values:

```bash
export DEFAULT_DOMAIN="${CLOUDRON_APP_DOMAIN}"
export IS_HTTPS_ENABLED="true"
# export GEOLITE_LICENSE_KEY= # Do not provide this instead of providing an empty value
export AUTO_RESOLVE_TITLES="false"
export DEFAULT_SHORT_CODES_LENGTH="5"
#export DELETE_SHORT_URL_THRESHOLD=
```

**Reply by msbt** [reply]
I've taken a different approach now and it works. I'm still not sure what happened though, one of the main differences in the images is the PHP Version, maybe that's something to check out when I have more time:
works: `PHP 8.1.15`
doesn't work: `PHP 8.1.2`

**Reply by acelaya** [reply]
I also thought that might be a bug on PHP itself. The other user who reported this was also using PHP 8.1.2, so it fits.

However I run it under all PHP 8.1 patches and couldn't reproduce it, so I discarded that reason.

Thanks for investigating it anyway 🙂

**Comment by acelaya** [comment, marked as answer]
The original discussion diverted a bit, so just providing an answer here.

Shlink now has an `init` command that can be used to set up an instance where configuration is provided via env vars, without having to go through the interactive installation process.

See more information on the docs https://shlink.io/documentation/command-line-interface/installation-tool/#init
